✨(cli) add a `pause` and `resume`  command to arnold CLI

madmatah opened the pull request:
## Purpose

We would like to be able to scale down an entire kubernetes namespace,
in order to save CPU and RAM resources on our development cluster.

## Proposal

This PR introduces 2 new commands in arnold CLI :

 - The `pause` command scales down all deployments, statefulsets and
   replication controllers that belongs to a namespace. Their previous
   replicas count is stored in an annotation.

 - The `resume` command does exactly the opposite, it restores the
   namespace in its previous state.

## How to test it

- `make cluster`
- `ANSIBLE_VAULT_PASSWORD=arnold bin/arnold -a hello -c eugene -d -e development bootstrap`
- `kubectl get deployment`
- `bin/arnold -c eugene -e development pause`
- `kubectl get deployment`
- `bin/arnold -c eugene -e development resume`

madmatah committed:
✨(cli) add `pause` and `resume` commands to arnold CLI

We would like to be able to scale down an entire kubernetes namespace,
in order to save CPU and RAM resources on our development cluster.
This PR introduces 2 new commands in arnold CLI :
 - The `pause` command scales down all deployments, statefulsets and
   replication controllers that belongs to a namespace. Their previous
   replicas count is stored in an annotation.
 - The `resume` command does exactly the opposite, it restores the
   namespace in its previous state.

jmaupetit commented on bin/arnold:
```diff
@@ -210,9 +210,11 @@ COMMANDS:
+  ingresses             create ingresses and redirections (if any)
   init                  initialize a namespace with all its volumes and configmaps
+  pause                 scale down all deployments of the namespace
+  resume                scale up all deployments of the namespace
```
```suggestion
  resume                restore all deployments of the namespace
```

jmaupetit commented on bin/arnold:
```diff
@@ -547,6 +549,68 @@ function _set_tray_vars() {
+
+    log info "Scaling down ${entity_type} ${entity_name}"
+
+    # Annotation the entity to store the current number of replicas
```
```suggestion
    # Annotate the entity to store the current number of replicas
```

jmaupetit commented on bin/arnold:
```diff
@@ -547,6 +549,68 @@ function _set_tray_vars() {
+  do
+    local entity_name="${entity_status[0]}"
+    local entity_replicas="${entity_status[1]}"
+    local entity_old_replicas_annotation="${entity_status[2]}"
```
Why not declare these variables in the previous block and only set them here?

jmaupetit commented on bin/arnold:
```diff
@@ -547,6 +549,68 @@ function _set_tray_vars() {
+  done
+}
+
+# _k8s_resume_entity : scale down all kubernetes entities of a specific type
```
```suggestion
# _k8s_resume_entity : restore all kubernetes entities of a specific type
```

jmaupetit commented on bin/arnold:
```diff
@@ -547,6 +549,68 @@ function _set_tray_vars() {
+  do
+    local entity_name="${entity_status[0]}"
+    local entity_replicas="${entity_status[1]}"
+    local entity_old_replicas_annotation="${entity_status[2]}"
```
ditto?

jmaupetit commented on bin/arnold:
```diff
@@ -547,6 +549,68 @@ function _set_tray_vars() {
+
+    # We only process the entities that are paused
+    if [[ "${entity_old_replicas_annotation}" = "NOT-PAUSED" ]]; then
+      log info "Ignoring ${entity_type} ${entity_name} that is already resumed"
```
```suggestion
      log info "Ignoring ${entity_type} ${entity_name} that is not paused"
```

jmaupetit commented on bin/arnold:
```diff
@@ -824,6 +888,35 @@ function reset_vault_pw() {
+  log info "Done"
+}
+
+# Scale up all deployments of the namespace that have been
```
```suggestion
# Restore all deployments of the namespace that have been
```

jmaupetit reviewed (approved):
I only have minor suggestions. It's brilliant. :clap: 

![](https://media4.giphy.com/media/l0He0cVv8lGggpruo/giphy.gif)

madmatah commented:
> I only have minor suggestions. It's brilliant. 

@jmaupetit Thanks! I addressed all your feedbacks.

FYI, I also changed the commit message title from:

> add a `pause` and `resume` command to arnold CLI

to:

> add `pause` and `resume` commands to arnold CLI
